### PR TITLE
evo2-sae: probing primitives + harness (metrics, ActivationBuffer, probe CLI)

### PR DESCRIPTION
**Pull request by polinabinder1**
## Summary
**SAE probing primitives + harness** for Evo2 — scoring metrics + `ActivationBuffer` + per-feature **annotation persistence**, all pure functions of codes + labels. Lives in the evo2 recipe at **`evo2_sae.eval.probing`** — kept out of the shared `sae` library because it's evo2-specific (the shared `sae.eval` keeps `loss_recovered` / `sparsity` / `dead_latents` for esm2/codonfm).

**Stacked on #1622** (uses the `evo2_sae` package). #1630 supplies the eval labels.

> **Stack update (2026-06-29):** #1636 (probing harness + per-domain label producers) was merged into this branch — its code (`evo2_buffer.py`, `labelers.py`, `annot_tracks.py`, `euk_windows.py`, `probe.py`, `probe_loss_recovered.py`, their tests) is now part of this PR. Scope is **probing primitives + harness + annotation persistence**.

## Contents — `evo2_sae.eval.probing`
**Primitives** (pure functions of codes + labels)
- `ActivationBuffer` (codes + optional dense twin + per-token labels + per-concept instance ids)
- AUROC: `auroc_all`, `auroc_vec`, `best_single_train_test`
- decoders: `fit_logreg` / `fit_softmax` / `macro_auroc` / `decode_eval`
- `domain_f1` (precision-per-nt, recall-per-instance)
- `annotate_features` — per-feature best concept by **AUROC** → the annotation table
- `annotate_features_domain` — per-feature best region concept by **domain-F1** → the annotation table (the interval sibling of `annotate_features`)

**Probing harness + label producers**
- `evo2_buffer.py` — DNA → `ActivationBuffer` (the one model-touching step: Evo2 → layer-L residual → `SAE.encode` + per-token labels)
- `labelers.py` / `annot_tracks.py` / `euk_windows.py` — per-domain biological labelers + BED/GFF interval tracks + euk gene-structure windows
- `probe.py` — the probing CLI (`extract` / `auroc` / `annotate` / `linear` / `euk-f1` / `domain-eval`)
- `probe_loss_recovered.py` — SAE **fidelity** via the shared `sae.eval.loss_recovered` (companion diagnostic; independent of the annotation flow)

## Annotation persistence — AUROC + domain-F1 → one parquet
Two label shapes, one output. **Point** concepts (start codon, splice site, …) are scored by AUROC on the continuous activations; **region** concepts (exon/intron/CDS, or any BED/GFF track) by instance-level domain-F1. `annotate` writes both into the single feature-annotation parquet the engine/dashboard load.

- **`annotate` auto-detects the buffer.** Point buffers (dense twin, no instance ids) → AUROC; region buffers (instance ids, no dense) → domain-F1. No flag to remember.
- **`euk-f1` / `domain-eval` gain `--out`** to save their region `ActivationBuffer` (no dense twin; per-concept instance ids, with the euk `cds` mask re-keyed to `gene` instances) — so the region path splits into the same *encode once → score offline* shape the point path already has.
- **`annotate` takes multiple `--acts`** and merges point + region buffers into one parquet, deduped per feature: **AUROC wins ties, domain-F1 fills the features AUROC left unlabeled** (the two scores are on different scales, so they're not compared numerically). New `method` / `score` columns record provenance; `feature_id` / `label` / `max_activation` are preserved for the engine contract.

## How to use
```python
from evo2_sae.eval.probing import auroc_all, annotate_features, annotate_features_domain
au  = auroc_all(codes, labels)                                    # [F, L]
ann = annotate_features(codes, labels, names, min_auroc=0.85)     # [{feature_id, label, auroc}]
```
End-to-end CLI ( `$L` = the SAE's layer — nothing in the code path is layer-specific ):
```bash
# 1. point buffer (needs the model): SAE codes + dense twin + per-token labels
probe extract     --evo2-ckpt-dir $CKPT --sae-checkpoint $SAE --layer $L --fasta probe.fa --out point.npz
# 2. region buffer from any BED/GFF tracks (needs the model)
probe domain-eval --evo2-ckpt-dir $CKPT --sae-checkpoint $SAE --layer $L --fasta seqs.fa \
      --track exon=refseq.gff3:exon --track cCRE=encode.bed --out region.npz
# 3. persist annotations (no model): AUROC for point, domain-F1 for region, merged into one parquet
probe annotate --acts point.npz region.npz --out feature_annotations.parquet
```

## Design decisions (the non-trivial ones)
- **Scoring is a pure function of a saved buffer — the model-touching step is isolated to one file.** `ActivationBuffer` (codes + optional dense twin + per-token labels + instance ids) is the contract; *all* metrics are pure CPU functions of it. Only `evo2_buffer.forward_codes` touches the model (Evo2 → layer-L → `SAE.encode`). So the expensive GPU pass runs once and every metric / re-analysis afterward runs on a laptop from the `.npz` — no model, fully reproducible. `annotate` itself never loads a model.
- **The buffer carries a dense-residual "twin."** Keeping the raw layer-L residual alongside the SAE codes lets the *same* artifact answer "does sparsification cost anything?" — SAE-vs-dense single- and multi-feature probes, no re-encode. Region buffers drop the twin (they don't run the linear probe) and instead carry per-concept instance ids for domain-F1's recall term.
- **One `annotate`, two metrics, one parquet.** Point vs region is inferred from the buffer (instances + no dense ⇒ region), so the persistence path is uniform and the dashboard reads a single file. AUROC-wins-then-F1-fills avoids comparing scores that live on different scales (AUROC ∈ [0.5,1] threshold-free; domain-F1 ∈ [0,1] thresholded).
- **Metrics are model-agnostic but live in the recipe package** (`evo2_sae.eval.probing`), beside `loss_recovered`/`reconstruction`/`sparsity`/`dead_latents`. They're the newest and evo2-driven; the established shared metrics stay in `sae.eval`. (Folding the probing primitives up into shared `sae` is the planned follow-up — they're written model-agnostic precisely so that's a move, not a rewrite.)
- **Hand-rolled, GPU-vectorized over the full feature dictionary** — `auroc_all` is one `[features × concepts]` pass; sklearn would be a per-`(feature, concept)` CPU loop over tens of thousands of features, and `fit_logreg` avoids the `sklearn.LogisticRegression` scaling wall that forced CodonFM to subsample.
- **Winner's-curse correction** — `best_single_train_test` selects the best single feature on **train** and reports *its* AUROC on **test**, so a concept's "best single feature" score isn't the optimistic max-over-features on the same data.
- **`domain_f1` is instance-level** (precision-per-nt, recall-per-annotation) and read **against a shuffle null** — for sparse regional concepts (exon/CDS) where per-token AUROC undersells, with the null guarding against high-prevalence concepts looking strong for free (intron ≈ 1.09× null = degenerate).
- **`probe_loss_recovered` reuses the shared `sae.eval.loss_recovered`** — SAE fidelity stays in the shared lib (esm2/codonfm use it too); only the Evo2 callables (a layer forward-hook + teacher-forced CE + denormalized reconstruction) are local. No duplicate metric. It's a companion diagnostic — orthogonal to the annotation flow.
- **Optional second GPU for the build step (`--auroc-device`).** Building a buffer holds two big things at once: the pinned Evo2 model (Megatron won't free it in-process) *and* the multi-GB code/dense buffer + the vectorized AUROC matmul over the whole dictionary. `--auroc-device` lets that buffer/scoring live on a **second** GPU so it doesn't fight the pinned model for memory. It's **optional** (defaults to `--device`) and matters **only for the in-process build-and-score path** (`extract` / `euk-f1` / `domain-eval`): once the buffer is saved, scoring is the pure-function path — no model, CPU or any GPU. A peak-memory accommodation for one step on large models, **not** a standing two-GPU requirement.

## Tests
No dedicated CI lane (deferred — see #1622). Run via the recipe:
```bash
cd interpretability/sparse_autoencoders/recipes/evo2
bash .ci_build.sh && source .ci_test_env.sh        # or: PYTHONPATH=src:../../sae/src
pytest tests/test_probing.py tests/test_probe_integration.py tests/test_labelers.py tests/test_annot_tracks.py tests/test_euk_windows.py
```
- **Primitives — CPU, no model:** AUROC vs a pairwise oracle, `domain_f1` vs a hand-computed reference, `best_single` winner's-curse flip, `decode_eval` separability, `annotate_features` best-concept, buffer roundtrip (incl. dense twin + instance ids), tie-correct ranks, degenerate-label / tie / sparse edge cases, `standardize` zero-variance floor.
- **Harness — CPU, no model:** label producers (`labelers` / `annot_tracks` / `euk_windows`) + the probe-CLI seam — buffer save/load, AUROC/annotate/linear over a planted feature, **domain-F1 annotate on a region buffer**, **point+region merge (AUROC wins)**, and the **`_region_buffer` helper** (mask stacking + cds→gene re-keying). The real-engine buffer / loss-recovered path is gated by `@pytest.mark.skipif(not torch.cuda.is_available())`.

## Why hand-rolled (not sklearn / torchmetrics) — checked, not a win
GPU-vectorized over the **whole feature dictionary** in one pass; the library options are CPU and per-`(scores, label)`, so a tens-of-thousands-feature dictionary becomes an equally large CPU loop. Function by function:
- **`auroc_all`** — no library computes a vectorized `[features × labels]` AUROC matrix on GPU. Kept.
- **`domain_f1`, `best_single_train_test`, `annotate_features`, `annotate_features_domain`** — bespoke (instance-F1, winner's-curse, per-feature assignment); no library equivalent.
- **`fit_logreg` / `fit_softmax` / `decode_eval`** — the only sklearn-replaceable code, but they fit on the full `[N, F]` SAE-code matrix, exactly where CodonFM hit the `sklearn.LogisticRegression` scaling wall and had to subsample. Swapping reintroduces that coverage loss + a runtime dep. Net regression.
- **`ActivationBuffer` / `split_indices` / `standardize`** — `np.savez` + tiny helpers; nothing to gain.

Conclusion: the module stays `torch` + `numpy`-only. Each metric is a standard formula (Mann–Whitney rank-AUROC, Adam BCE/softmax, instance-F1) vectorized for full-dictionary GPU scale, and each is validated against an independent reference in the tests.

### CODE REVIEW

**Comment by polinabinder1** [comment]
@coderabbitai review

**Comment by polinabinder1** [comment]
Addressed the two nitpicks in `57837ec7`: documented the `allow_pickle=True` trust caveat on `ActivationBuffer.load`, and added a comment explaining the `+2` remap-tensor sizing (index-by-max-id + sentinel headroom). Tests still green (6 passed).

**Review by jwilber — APPROVED** [review]
